sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java

```
@@ -0,0 +1,194 @@
+    try (Jedis connection = connect()) {
+      String[] infos = connection.info().split("\n");
+      for (String info : infos) {
+        if (info.startsWith("cluster_enabled:")) {
+          if (info.substring("cluster_enabled:".length()).contains("1")) {
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+  }
+
+  /**
+   * Populate the display data with connectionConfiguration details.
+   */
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder.addIfNotNull(DisplayData.item("host", host()));
+    builder.addIfNotNull(DisplayData.item("port", port()));
+    builder.addIfNotNull(DisplayData.item("timeout", timeout()));
+  }
+
+  public List<RedisService.RedisNode> getNodes() {
+    if (this.isClusterEnabled()) {
+      return getClusterNodes();
+    }
+    return Collections.emptyList();
+  }
+
+  public List<RedisService.RedisNode> getClusterNodes() {
+    try (Jedis jedis = connect()) {
+      // cluster slots is an array containing for each slot
+      // - start slot range
+      // - end slot range
+      // - master for slot range represented as nested IP/port array
+      // - first replica of master for slot range
+      // - second replica
+      // - ...continues until all replicas for this master are returned
+      // See https://redis.io/commands/cluster-slots for details
+      List<Object> slots = jedis.clusterSlots();
+
+      for (Object slotInfoObj : slots) {
+        List<Object> slotInfo = (List<Object>) slotInfoObj;
+
+        int startSlot = Integer.parseInt((String) slotInfo.get(0));
+        int endSlot = Integer.parseInt((String) slotInfo.get(1));
+
+        if (slotInfo.size() <= MASTER_NODE_INDEX) {
+          continue;
+        }
+
+        // host infos
+        int size = slotInfo.size();
+        for (int i = MASTER_NODE_INDEX; i < size; i++) {
+          List<Object> hostInfos = (List<Object>) slotInfo.get(i);
+          if (hostInfos.size() <= 0) {
```

You actually need it to have size exactly 2, right? And if it's not 2, then you'd probably want to throw an exception rather than silently skip it?